### PR TITLE
Improve Gemma4 MTP server batching

### PR DESCRIPTION
**Pull request by Blaizzy**
## Summary

 Merge order #1166, #1167, #1169

- Improves Gemma4 MTP speculative decoding in the server path.
- Adds a small speculative request coalescing window so concurrent requests form a real MTP batch before decode starts.
- Uses the CLI-style unbatched MTP loop for singleton server requests while keeping the batched loop for multi-request batches.
- Fixes batched/singleton MTP cache and mask handling around shared KV state metadata and one-row array offsets.
- Adds focused tests for the server MTP dispatch, cache selection, batch coalescing, and Gemma4 assistant mask behavior.

## Results

Benchmarks used `/v1/chat/completions`, prompt `Make a program to find pi`, `temperature=0`, `max_tokens=1200`, and MTP `--draft-block-size 3`.

| Model | Baseline single | MTP single | Single speedup | Baseline batch=4 | MTP batch=4 | Batch=4 speedup |
|---|---:|---:|---:|---:|---:|---:|
| `gemma-4-E2B-it-5bit` | 135.630 | 171.267 | 1.26x | 378.128 | 383.453 | 1.01x |
| `gemma-4-E4B-it-5bit` | 94.697 | 121.513 | 1.28x | 225.581 | 257.403 | 1.14x |
| `gemma-4-26B-A4B-it-5bit` | 84.705 | 112.325 | 1.33x | 212.159 | 193.837 | 0.91x |
| `gemma-4-31B-it-5bit` | 24.854 | 33.774 | 1.36x | 44.757 | 61.626 | 1.38x |

Higher-concurrency server batch results with the same block size:

| Model | Batch | Baseline tok/s | MTP tok/s | Speedup |
|---|---:|---:|---:|---:|
| `gemma-4-E2B-it-5bit` | 8 | 513.203 | 643.447 | 1.25x |
| `gemma-4-E4B-it-5bit` | 8 | 292.623 | 462.836 | 1.58x |
| `gemma-4-26B-A4B-it-5bit` | 8 | 281.583 | 289.111 | 1.03x |
| `gemma-4-31B-it-5bit` | 6 | 49.336 | 81.792 | 1.66x |

31B batch=8 did not complete in this environment, so batch=6 is reported for 31B.

Screenshot 2026-05-09 at 23 38 26

## Validation

- `uv run pytest mlx_vlm/tests/test_server.py mlx_vlm/tests/test_speculative.py -q`
- Result: `108 passed, 5 warnings`

### CODE REVIEW

File: mlx_vlm/speculative/drafters/gemma4_assistant/masked_embedder.py [hunk 1]
```diff
@@ -46,7 +46,35 @@ def __call__(self, hidden_states: mx.array, lm_head_weight: mx.array) -> mx.arra
...
+    def argmax(self, hidden_states: mx.array, lm_head_weight: mx.array) -> mx.array:
+        """Return greedy tokens without materializing full-vocab logits."""
+        selected_canonical, selected_logits = self._selected_logits(
+            hidden_states, lm_head_weight
+        )
+        best = mx.argmax(selected_logits, axis=-1)[..., None]
+        selected_canonical = selected_canonical.reshape(*hidden_states.shape[:2], -1)
+        return mx.take_along_axis(selected_canonical, best, axis=-1).squeeze(-1)
 
+    def _selected_logits(
```
Comment (lucasnewman) [inline mlx_vlm/speculative/drafters/gemma4_assistant/masked_embedder.py:74]: Maybe a candidate for compilation?

**Review by lucasnewman — APPROVED** [review]
LGTM
